File: wakeonlan/src/test/java/com/github/shingyx/wakeonlan/MagicPacketProcessorTest.java
package com.github.shingyx.wakeonlan;

import org.junit.Test;

import static org.junit.Assert.*;

public class MagicPacketProcessorTest {

    @Test
    public void convertMacAddressString_allZero() throws Exception {
        runHappyTest(new byte[]{0, 0, 0, 0, 0, 0}, "00:00:00:00:00:00");
    }

    @Test
    public void convertMacAddressString_allF() throws Exception {
        runHappyTest(new int[]{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}, "ff:ff:ff:ff:ff:ff");
    }

    @Test
    public void convertMacAddressString_aToF() throws Exception {
        runHappyTest(new int[]{0xab, 0xcd, 0xef, 0xfe, 0xdc, 0xba}, "ab:cd:ef:fe:dc:ba");
    }

    @Test
    public void convertMacAddressString_upperCase() throws Exception {
        runHappyTest(new int[]{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}, "FF:FF:FF:FF:FF:FF");
    }

    @Test
    public void convertMacAddressString_hyphenSeparator() throws Exception {
        runHappyTest(new byte[]{0, 0, 0, 0, 0, 0}, "00-00-00-00-00-00");
    }

    @Test
    public void convertMacAddressString_containsInvalidHex() throws Exception {
        runSadTest("00:00:00:00:00:0g");
    }

    @Test
    public void convertMacAddressString_tooShort() throws Exception {
        runSadTest("ff:ff:ff:ff:ff:f");
    }

    @Test
    public void convertMacAddressString_tooLong() throws Exception {
        runSadTest("ff:ff:ff:ff:ff:fff");
    }

    @Test
    public void convertMacAddressString_tooLong2() throws Exception {
        runSadTest("ff:ff:ff:ff:ff:ff-");
    }

    @Test
    public void convertMacAddressString_doubleSeparator() throws Exception {
        runSadTest("ff:ff:ff:ff:ff::f");
    }

    @Test
    public void convertMacAddressString_noSeparators() throws Exception {
        runSadTest("ffffffffffff");
    }

    @Test
    public void convertMacAddressString_emptyString() throws Exception {
        runSadTest("");
    }

    @Test
    public void convertMacAddressString_null() throws Exception {
        runSadTest(null);
    }

    private void runHappyTest(byte[] expected, String input) {
        assertArrayEquals(expected, MagicPacketProcessor.convertMacAddressString(input));
    }

    private void runHappyTest(int[] expected, String input) {
        byte[] expectedBytes = new byte[expected.length];
        for (int i = 0; i < expected.length; i++) {
            expectedBytes[i] = (byte) expected[i];
        }
        runHappyTest(expectedBytes, input);
    }

    private void runSadTest(String input) {
        try {
            MagicPacketProcessor.convertMacAddressString(input);
            fail("Expected an exception");
        } catch (IllegalArgumentException e) {
            // Caught an IllegalArgumentException
        }
    }
}

File: wakeonlan/src/test/java/com/github/shingyx/wakeonlan/MagicPacketProcessorTest.kt
package com.github.shingyx.wakeonlan

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.fail
import org.junit.Test

class MagicPacketProcessorTest {

    @Test
    fun convertMacAddressString_allZero() {
        runHappyTest(intArrayOf(0, 0, 0, 0, 0, 0), "00:00:00:00:00:00")
    }

    @Test
    fun convertMacAddressString_allF() {
        runHappyTest(intArrayOf(0xff, 0xff, 0xff, 0xff, 0xff, 0xff), "ff:ff:ff:ff:ff:ff")
    }

    @Test
    fun convertMacAddressString_aToF() {
        runHappyTest(intArrayOf(0xab, 0xcd, 0xef, 0xfe, 0xdc, 0xba), "ab:cd:ef:fe:dc:ba")
    }

    @Test
    fun convertMacAddressString_upperCase() {
        runHappyTest(intArrayOf(0xff, 0xff, 0xff, 0xff, 0xff, 0xff), "FF:FF:FF:FF:FF:FF")
    }

    @Test
    fun convertMacAddressString_hyphenSeparator() {
        runHappyTest(intArrayOf(0, 0, 0, 0, 0, 0), "00-00-00-00-00-00")
    }

    @Test
    fun convertMacAddressString_containsInvalidHex() {
        runSadTest("00:00:00:00:00:0g")
    }

    @Test
    fun convertMacAddressString_tooShort() {
        runSadTest("ff:ff:ff:ff:ff:f")
    }

    @Test
    fun convertMacAddressString_tooLong() {
        runSadTest("ff:ff:ff:ff:ff:fff")
    }

    @Test
    fun convertMacAddressString_tooLong2() {
        runSadTest("ff:ff:ff:ff:ff:ff-")
    }

    @Test
    fun convertMacAddressString_doubleSeparator() {
        runSadTest("ff:ff:ff:ff:ff::f")
    }

    @Test
    fun convertMacAddressString_noSeparators() {
        runSadTest("ffffffffffff")
    }

    @Test
    fun convertMacAddressString_emptyString() {
        runSadTest("")
    }

    private fun runHappyTest(expected: IntArray, input: String) {
        val expectedBytes = expected.map { it.toByte() }.toByteArray()
        assertArrayEquals(expectedBytes, convertMacAddressString(input))
    }

    private fun runSadTest(input: String) {
        try {
            convertMacAddressString(input)
            fail("Expected an exception")
        } catch (e: IllegalArgumentException) {
            // Caught an IllegalArgumentException
        }
    }
}
